File: 프로그래머스/unrated/176963. 추억 점수/추억 점수.java
import java.util.HashMap;

class Solution {
    public int[] solution(String[] name, int[] yearning, String[][] photo) {
        int[] scores = new int[photo.length];
        HashMap<String, Integer> map = new HashMap<>();
        
        for(int i = 0; i < name.length; i++) map.put(name[i], yearning[i]);
        for(int i = 0; i < photo.length; i++) {
            int score = 0;
            String[] pho = photo[i];
            for(String p : pho) {
                if(map.get(p) != null) score += map.get(p);
            }
            scores[i] = score;
        }
        
        return scores;
    }
} 

File: 프로그래머스/1/176963. 추억 점수/추억 점수.js
function solution(names, scores, photos) {
    const result = [];
    const map = new Map();
    
    initScore(map, names, scores);
    for (photo of photos) {
        result.push(calculateTotalScore(map, photo));
    }
    
    return result;
}

const initScore = (map, names, scores) => {
    for (let i = 0; i < names.length; i++) {
        const name = names[i];
        const score = scores[i];
        map.set(name, score);
    }
    
}

const calculateTotalScore = (map, photo) => {
    let totalScore = 0;
    
    for (const name of photo) {
        const score = map.get(name);
        if (score) {
            totalScore += score;
        }
    }
    
    return totalScore;
}